File: javascript/routes/injection_sql.js
/**
 * SQL Injection — OWASP 2021 A03 Injection | OWASP 2025 A05 Injection.
 *
 * Reference module: the comment/annotation format used here is mirrored by
 * every other vulnerability module in this app. Each permutation carries a
 * standard header describing the flaw, its CWE, an exploitation-likelihood
 * rating, an example payload, and the corresponding safe pattern.
 */
'use strict';

const express = require('express');
const { getDb } = require('../db');

const router = express.Router();

// ===========================================================================
// PERMUTATION 1 — SQL injection via template-literal concatenation (auth bypass)
// OWASP 2021: A03 Injection      OWASP 2025: A05 Injection
// CWE-89: Improper Neutralization of Special Elements used in an SQL Command
// EXPLOITATION LIKELIHOOD: HIGH
//   Reachable pre-authentication, deterministic, no tooling required.
//   The textbook  ' OR '1'='1  / username=admin'--  payloads bypass login.
// PREVALENCE TODAY: greenfield LOW-MED (modern ORMs parameterize by default;
//   recurs in raw SQL and AI-generated snippets) | legacy/10yr tech-debt HIGH.
// EXAMPLE:
//   GET /injection/sql/login?username=admin'--&password=whatever
//   GET /injection/sql/login?username=x' OR '1'='1&password=x
// FIX: parameterize (see /injection/sql/login-safe).
// ===========================================================================
router.get('/login', (req, res) => {
  const username = req.query.username || '';
  const password = req.query.password || '';
  // VULNERABLE: user input concatenated straight into the SQL string.
  const sql =
    `SELECT id, username, role FROM users ` +
    `WHERE username = '${username}' AND password = '${password}'`;
  try {
    const row = getDb().prepare(sql).get();
    res.json({ authenticated: !!row, as: row || null, query: sql });
  } catch (err) {
    res.status(500).json({ query: sql, error: err.message });
  }
});

// ===========================================================================
// PERMUTATION 2 — SQL injection via string concatenation (UNION data theft)
// OWASP 2021: A03 Injection      OWASP 2025: A05 Injection
// CWE-89
// EXPLOITATION LIKELIHOOD: HIGH
//   A UNION SELECT reads arbitrary columns/tables (e.g. dump password hashes
//   and SSNs from the users table) through a search box.
// PREVALENCE TODAY: greenfield MEDIUM (raw SQL persists in search/reporting/
//   analytics paths) | legacy/10yr tech-debt HIGH.
// EXAMPLE:
//   GET /injection/sql/search?q=zzz' UNION SELECT username, password, ssn FROM users--
// FIX: parameterize the WHERE clause.
// ===========================================================================
router.get('/search', (req, res) => {
  const q = req.query.q || '';
  // VULNERABLE: untrusted input concatenated into a LIKE clause.
  const sql = `SELECT title, body, owner FROM notes WHERE body LIKE '%${q}%'`;
  try {
    const results = getDb().prepare(sql).all();
    res.json({ query: sql, results });
  } catch (err) {
    res.status(500).json({ query: sql, error: err.message });
  }
});

// ===========================================================================
// PERMUTATION 3 — SQL injection via a dynamic ORDER BY clause
// OWASP 2021: A03 Injection      OWASP 2025: A05 Injection
// CWE-89
// EXPLOITATION LIKELIHOOD: MEDIUM
//   Column/keyword positions cannot be bound as parameters, so developers
//   often concatenate sort fields. Exploitable for blind boolean/error-based
//   extraction; needs more skill than a UNION.
// PREVALENCE TODAY: greenfield MED-HIGH (SQL identifiers/sort columns cannot be
//   bound, so concatenated ORDER BY stays common in modern ORM apps) | legacy HIGH.
// EXAMPLE:
//   GET /injection/sql/notes?sort=body--
//   GET /injection/sql/notes?sort=(CASE WHEN (SELECT 1)=1 THEN title ELSE body END)
// FIX: allow-list sortable column names.
// ===========================================================================
router.get('/notes', (req, res) => {
  const sort = req.query.sort || 'id';
  // VULNERABLE: sort field concatenated into ORDER BY.
  const sql = `SELECT id, owner, title FROM notes ORDER BY ${sort}`;
  try {
    const results = getDb().prepare(sql).all();
    res.json({ query: sql, results });
  } catch (err) {
    res.status(500).json({ query: sql, error: err.message });
  }
});

// ===========================================================================
// SAFE REFERENCE — parameterized query. Shown so the vulnerable/safe pair can
// be diffed by SAST tooling and learners.
// ===========================================================================
router.get('/login-safe', (req, res) => {
  const username = req.query.username || '';
  const password = req.query.password || '';
  const row = getDb()
    .prepare('SELECT id, username, role FROM users WHERE username = ? AND password = ?')
    .get(username, password);
  res.json({ authenticated: !!row, as: row || null });
});

module.exports = { base: '/injection/sql', router };

File: go/injection_sql.go
// SQL Injection — OWASP 2021 A03 Injection | OWASP 2025 A05 Injection.
//
// Reference module for the Go app: mirror this annotation format and the
// init()-based registration in every other Go vulnerability module.
package main

import (
	"fmt"
	"net/http"
)

func init() {
	register("GET /injection/sql/login", "SQLi: auth bypass via string concat", sqlLogin)
	register("GET /injection/sql/search", "SQLi: UNION data theft", sqlSearch)
	register("GET /injection/sql/notes", "SQLi: dynamic ORDER BY", sqlNotes)
	register("GET /injection/sql/login-safe", "SAFE: parameterized query", sqlLoginSafe)
}

// ============================================================================
// PERMUTATION 1 — SQL injection via string concatenation (auth bypass)
// OWASP 2021 A03 Injection      OWASP 2025 A05 Injection
// CWE-89: Improper Neutralization of Special Elements used in an SQL Command
// EXPLOITATION LIKELIHOOD: HIGH — pre-auth, deterministic, no tooling; the
//   textbook  admin'--  /  ' OR '1'='1  payloads bypass the login.
// PREVALENCE TODAY: greenfield LOW-MED (database/sql encourages placeholders;
//   recurs when devs fmt.Sprintf queries) | legacy/10yr tech-debt HIGH.
// EXAMPLE: GET /injection/sql/login?username=admin'--&password=x
// FIX: use parameter placeholders (see /injection/sql/login-safe).
// ============================================================================
func sqlLogin(w http.ResponseWriter, r *http.Request) {
	u := r.URL.Query().Get("username")
	p := r.URL.Query().Get("password")
	// VULNERABLE: user input concatenated straight into the SQL string.
	q := fmt.Sprintf("SELECT id, username, role FROM users WHERE username='%s' AND password='%s'", u, p)
	rows, err := db.Query(q)
	if err != nil {
		writeJSON(w, 500, map[string]any{"query": q, "error": err.Error()})
		return
	}
	defer rows.Close()
	list, err := queryToMaps(rows)
	if err != nil {
		writeJSON(w, 500, map[string]any{"query": q, "error": err.Error()})
		return
	}
	writeJSON(w, 200, map[string]any{"query": q, "authenticated": len(list) > 0, "as": first(list)})
}

// ============================================================================
// PERMUTATION 2 — SQL injection via fmt.Sprintf (UNION data theft)
// OWASP 2021 A03 Injection      OWASP 2025 A05 Injection
// CWE-89
// EXPLOITATION LIKELIHOOD: HIGH — a UNION SELECT dumps password hashes and SSNs
//   from the users table through a notes search box.
// PREVALENCE TODAY: greenfield MEDIUM (raw SQL persists in search/reporting) |
//   legacy/10yr tech-debt HIGH.
// EXAMPLE: GET /injection/sql/search?q=zzz' UNION SELECT username, password, ssn FROM users--
// FIX: parameterize the WHERE clause.
// ============================================================================
func sqlSearch(w http.ResponseWriter, r *http.Request) {
	term := r.URL.Query().Get("q")
	// VULNERABLE: interpolated LIKE clause.
	q := fmt.Sprintf("SELECT title, body, owner FROM notes WHERE body LIKE '%%%s%%'", term)
	rows, err := db.Query(q)
	if err != nil {
		writeJSON(w, 500, map[string]any{"query": q, "error": err.Error()})
		return
	}
	defer rows.Close()
	list, err := queryToMaps(rows)
	if err != nil {
		writeJSON(w, 500, map[string]any{"query": q, "error": err.Error()})
		return
	}
	writeJSON(w, 200, map[string]any{"query": q, "results": list})
}

// ============================================================================
// PERMUTATION 3 — SQL injection via a dynamic ORDER BY clause
// OWASP 2021 A03 Injection      OWASP 2025 A05 Injection
// CWE-89
// EXPLOITATION LIKELIHOOD: MEDIUM — identifiers can't be placeholders, so a
//   concatenated sort field enables blind boolean/error-based extraction.
// PREVALENCE TODAY: greenfield MED-HIGH (ORDER BY columns can't be bound, so
//   concatenation stays common even in modern code) | legacy HIGH.
// EXAMPLE: GET /injection/sql/notes?sort=body--
// FIX: allow-list sortable column names.
// ============================================================================
func sqlNotes(w http.ResponseWriter, r *http.Request) {
	sort := r.URL.Query().Get("sort")
	if sort == "" {
		sort = "id"
	}
	// VULNERABLE: sort field concatenated into ORDER BY.
	q := fmt.Sprintf("SELECT id, owner, title FROM notes ORDER BY %s", sort)
	rows, err := db.Query(q)
	if err != nil {
		writeJSON(w, 500, map[string]any{"query": q, "error": err.Error()})
		return
	}
	defer rows.Close()
	list, _ := queryToMaps(rows)
	writeJSON(w, 200, map[string]any{"query": q, "results": list})
}

// ============================================================================
// SAFE REFERENCE — parameterized query. Shown so the vulnerable/safe pair can
// be diffed by SAST tooling and learners.
// ============================================================================
func sqlLoginSafe(w http.ResponseWriter, r *http.Request) {
	u := r.URL.Query().Get("username")
	p := r.URL.Query().Get("password")
	rows, err := db.Query("SELECT id, username, role FROM users WHERE username=? AND password=?", u, p)
	if err != nil {
		writeJSON(w, 500, map[string]any{"error": err.Error()})
		return
	}
	defer rows.Close()
	list, _ := queryToMaps(rows)
	writeJSON(w, 200, map[string]any{"authenticated": len(list) > 0, "as": first(list)})
}
